README
Sort-Naturally-XS version 0.8.0
==============================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2016 by Sergey Yurzin

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Sort::Naturally::XS.

0.1.0  Sat Oct 22 21:59:46 2016
	- original version; created by h2xs 1.23 with options
		-O -n Sort::Naturally::XS ./Sort-Naturally-XS/src/nsort.h

0.2.0  Wed Oct 26 22:40:00 2016
    - beta implemented; initial commit

0.3.0  Fri Oct 26 18:25:00 2016
    - support Sort::Naturally API, add ncmp subroutine

0.4.0 Sun Nov 13 22:17:00 2016
    - add nsort subroutine and minimal documentation

0.5.0 Thu Nov 17 11:21:00 2016
    - add light pre-process to comparator, increase performance

0.6.0 Wed Nov 23 22:08:00 2016
    - refactor pre-process in comparator, properly sort dimensions (like 2400x2600), improve tests

0.6.1 Sat Now 26 22:13:00 2016
    - support perl v5.10.1

0.6.2 Sat Now 26 22:40:00 2016
    - minor fix digit only sort, avoid useless iteration for each element

0.7.0 Tue Now 29 22:14:00 2016
    - add sorted subroutine with locale and reverse support

0.7.1 Thu Dec 01 21:38:00 2016
    - fix simultaneously locale and reverse sort

0.7.2 Wed Dec 07 23:13:00 2016
    - fix locale support, improve locale tests

0.7.3 Tue Dec 27 14:16:00 2016
    - fix perl 5.10.1 support

0.7.4 Wed Jan 25 22:13:00 2017
    - skip locale tests if detect not POSIX-conformant OS

0.7.5 Sun Jan 29 23:47:00 2017
    - fix make on BSD systems

0.7.6 Fri Feb 17 23:01:00 2017
    - improve portability, fix NetBSD support

0.7.7 Fri Jun 09 08:35:00 2017
    - update according to META CPAN

0.8.0 Tue Jan 31 22:08:00 2017
    - icu support implemented, cross-platform locale aware sort available


Makefile.PL
use 5.010001;
use strict;
use Config;
use ExtUtils::MakeMaker;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
	NAME         => 'Sort::Naturally::XS',
	VERSION_FROM => 'lib/Sort/Naturally/XS.pm', # finds $VERSION
	PREREQ_PM    => {}, # e.g., Module::Name => 1.1
	($] >= 5.005                                ? ## Add these new keywords supported since 5.005
		(ABSTRACT_FROM => 'lib/Sort/Naturally/XS.pm', # retrieve abstract from module
			AUTHOR     => 'Sergey Yurzin <jurzin.s@gmail.com>') : ()),
	LIBS         => is_mac_os() ? '-L/usr/local/opt/icu4c/lib' : '',
	DEFINE       => '', # e.g., '-DHAVE_SOMETHING'
	INC          => '-I. ' . (is_mac_os() ? '-I/usr/local/opt/icu4c/include' : ''),
	# Un-comment this if you add C files to link with later:
	# OBJECT            => '$(O_FILES)', # link all the C files too
	MYEXTLIB     => 'src/libnsort$(LIB_EXT)',
	OPTIMIZE     => '-O3',
	LDDLFLAGS    => $ExtUtils::MakeMaker::Config::Config{lddlflags}.' `pkg-config --libs --cflags icu-uc icu-io`',
	META_MERGE   => {
		'meta-spec' => {
			version => 2
		},
		resources   => {
			repository => {
				type => 'git',
				url  => 'https://github.com/CaballerosTeam/Sort-Naturally-XS.git',
				web  => 'https://github.com/CaballerosTeam/Sort-Naturally-XS',
			},
			bugtracker => {
				web    => 'https://github.com/CaballerosTeam/Sort-Naturally-XS/issues',
				mailto => 'jurzin.s@gmail.com',
			},
		},
	},
);
if (eval {
	require ExtUtils::Constant;
	1
}) {
	# If you edit these definitions to change the constants used by this module,
	# you will need to use the generated const-c.inc and const-xs.inc
	# files to replace their "fallback" counterparts before distributing your
	# changes.
	my @names = (qw());
	ExtUtils::Constant::WriteConstants(
		NAME         => 'Sort::Naturally::XS',
		NAMES        => \@names,
		DEFAULT_TYPE => 'IV',
		C_FILE       => 'const-c.inc',
		XS_FILE      => 'const-xs.inc',
	);

}
else {
	use File::Copy;
	use File::Spec;
	foreach my $file ('const-c.inc', 'const-xs.inc') {
		my $fallback = File::Spec->catfile( 'fallback', $file );
		copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
	}
}
sub MY::postamble {
'
$(MYEXTLIB): src/Makefile
		cd src && $(MAKE) $(PASSTHRU)
';
}

sub is_mac_os {
	return $Config{osname} eq 'darwin';
}

src/nsort.c
#include <stdlib.h>
#include <ctype.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdbool.h>
#include <unicode/ustring.h>
#include <unicode/ucol.h>
#include "nsort.h"

char *get_next_chunk(const char *, int *, bool *);
bool isDigit(const char);
bool isAlpha(const char);

int _ncmp(const char *a, const char *b, int reverse, UCollator *collator) {
    int len_a = strlen(a);
    int len_b = strlen(b);
    int offset_a = 0;
    int offset_b = 0;
    bool is_digit_a = NULL;
    bool is_digit_b = NULL;
    bool is_last_chunk_a_digit = NULL;
    bool is_last_chunk_b_digit = NULL;
    bool is_fist_char_a_alpha = isAlpha(a[0]);
    bool is_fist_char_b_alpha = isAlpha(b[0]);
    char *chunk_a;
    char *chunk_b;
    int chunk_a_int;
    int chunk_b_int;
    int result = 0;
    UChar u_chunk_a[100], u_chunk_b[100];
    UErrorCode errorCode = U_ZERO_ERROR;

    if (!collator && is_fist_char_a_alpha && is_fist_char_b_alpha) {
        offset_a = 1;
        offset_b = 1;
        result = (a[0] < b[0]) ? -1 : (a[0] > b[0]);
    } else if (is_fist_char_a_alpha && isDigit(b[0])) {
        result = 1;
    } else if (isDigit(a[0]) && is_fist_char_b_alpha) {
        result = -1;
    }

    if (result == 0) {
        while (offset_a != len_a && offset_b != len_b) {
            chunk_a = get_next_chunk(a, &offset_a, &is_digit_a);
            chunk_b = get_next_chunk(b, &offset_b, &is_digit_b);
            is_last_chunk_a_digit = !is_digit_a;
            is_last_chunk_b_digit = !is_digit_b;

            if (is_last_chunk_a_digit == is_last_chunk_b_digit) {
                if (is_last_chunk_a_digit) {
                    chunk_a_int = atoi(chunk_a);
                    chunk_b_int = atoi(chunk_b);
                    result = (chunk_a_int < chunk_b_int) ? -1 : (chunk_a_int > chunk_b_int);
                } else if (collator) {
                    u_strFromUTF8(u_chunk_a, 100, NULL, chunk_a, -1, &errorCode);
                    if (U_FAILURE(errorCode)) {
                        result = 0;
                    }
                    u_strFromUTF8(u_chunk_b, 100, NULL, chunk_b, -1, &errorCode);
                    if (U_FAILURE(errorCode)) {
                        result = 0;
                    }
                    UCollationResult collationResult = ucol_strcoll(collator, u_chunk_a, -1, u_chunk_b, -1);
                    result = (collationResult == UCOL_LESS) ? -1 : ((collationResult == UCOL_GREATER) ? 1 : 0);
                } else {
                    result = strcmp(chunk_a, chunk_b);
                }
            } else {
                if (is_last_chunk_a_digit) {
                    result = -1;
                } else {
                    result = 1;
                }
            }

            free(chunk_a);
            free(chunk_b);

            if (result != 0) {
                break;
            } else if (offset_a == len_a) {
                result = -1;
            } else if (offset_b == len_b) {
                result = 1;
            }
        }
    }

    return reverse ? (-1 * result) : result;
}

char *get_next_chunk(const char *raw, int *offset, bool *is_digit) {
    if (*offset == 0) {
        *is_digit = isDigit(raw[0]);
    }

    int i;
    int len = 0;
    int raw_len = strlen(raw);
    for (i = *offset; i < raw_len; i++) {
        bool c_is_digit = isDigit(raw[i]);

        if (c_is_digit != *is_digit) {
            *is_digit = c_is_digit;
            len = i - *offset;
            break;
        } else if (i == raw_len - 1) {
            len = raw_len - *offset;
            *is_digit = !*is_digit;
        }
    }

    char *chunk = malloc((len + 1) * sizeof(char));
    strncpy(chunk, raw + *offset, len);
    chunk[len] = '\0';
    *offset += len;

    return chunk;
}

bool isDigit(const char c) {
    return (c >= '0' && c <= '9');
}

bool isAlpha(const char c) {
    return ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'));
}


lib/Sort/Naturally/XS.pm
package Sort::Naturally::XS;

use 5.010001;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw/Exporter/;

our @EXPORT = qw/ncmp nsort/;

our @EXPORT_OK = qw/sorted/;

our $VERSION = '0.8.0';

require XSLoader;
XSLoader::load('Sort::Naturally::XS', $VERSION);

sub sorted {
    my ($ar, %kwargs) = @_;

    Carp::confess('Not an ARRAY ref') if (ref $ar ne 'ARRAY');

    my $ar_copy = [@{$ar}];
    my $reverse = $kwargs{reverse} ? 1 : 0;
    my $locale = $kwargs{locale} || '';

    _sorted($ar_copy, $reverse, $locale);

    return $ar_copy;
}

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__
=encoding utf8


=head1 NAME

Sort::Naturally::XS - Perl extension for human-friendly ("natural") sort order

=head1 SYNOPSIS

  use Sort::Naturally::XS;

  my @mixed_list = qw/test21 test20 test10 test11 test2 test1/;

  my @result = nsort(@mixed_list); # @result is: test1 test2 test10 test11 test20 test21

  @result = sort ncmp @mixed_list; # same, but use standard sort function

  @result = sort {ncmp($a, $b)} @mixed_list; # same as ncmp, but argument pass explicitly

=head1 DESCRIPTION

Natural sort order is an ordering of mixed strings (consist of characters and digits) in alphabetical order, except
that digital parts are ordered as numbers. Natural sorting can be considered as a replacement of a standard
machine-oriented alphabetical sorting, because it is more convenient for human understanding. For example,
the following list:

  test21 test20 test10 test11 test2 test1

after performing a standard machine-oriented alphabetical sorting, will be as follows:

  test1 test10 test11 test2 test20 test21

The sequence appears unnatural, because test10 and test11 come before test2. On the other hand, natural sorting gives a
human-friendly sequence:

  test1 test2 test10 test11 test20 test21

now test2 comes before test10 and test11.


=head1 METHODS

=over 4

=item C<ncmp(LEFT, RIGHT)>

Replacement of the C<cmp> standard operator. LEFT and RIGHT variables are presented for comparison. Returns 1 if LEFT
should come before RIGHT, -1 if RIGHT should come before LEFT and 0 if LEFT and RIGHT match.

  # sort @list naturally, support in latest perl versions
  my @result = sort ncmp @list;

  # same, but arguments pass explicitly
  @result = sort {ncmp($a, $b)} @list;

  # more complex example, sort ARRAY of HASH refs by key 'foo' in descending order
  @result = sort {ncmp($b->{foo}, $a->{foo})} @list;

=item C<nsort(LIST)>

In list context returns a LIST sorted copy.

  my @result = nsort(@list);

=item C<sorted(ARRAY_REF, KWARGS)>

Returns an ARRAY ref to a sorted list. First argument is an ARRAY ref to the source list, followed by keyword arguments,
such as C<reverse> and C<locale>. If C<reverse> is true the source list is sorted in reverse order. If C<locale> is
specified, the sorting will be performed according to the locale aware settings.

  use Sort::Naturally::XS qw/sorted/;

  my $result = sorted($list);

  $result = sorted($list, reverse => 1); # $list will be sorted in descending order

  # $list will be sorted according to en_US.utf8 locale
  $result = sorted($list, locale => 'en_US.utf8');

=back


=head1 LOCALE AWARE SORTING

By default the sort function sorts according to a standard C locale or, if a C<use locale> pragma is in effect,
according to OS settings, which can be changed with the help of the C<setlocale> function. The use of both C<use locale>
and C<setlocale> has no effect on C<ncmp> and C<nsort>. The following example demonstrates this behavior:

  use POSIX;
  use Sort::Naturally::XS;

  my @list = ('a.'.'c', 'A'..'B');

  my @result_std = sort @list;
  my @result_ncmp = sort {ncmp($a, $b)} @list;
  # @result_std contains  A, B, C, a, b, c
  # @result_ncmp contains A, B, C, a, b, c

  use locale;
  # assumed that current locale is en_US.utf8
  @result_std = sort @list;
  @result_ncmp = sort {ncmp($a, $b)} @list;
  # @result_std contains  a, A, b, B, c, C
  # @result_ncmp contains A, B, C, a, b, c

  setlocale(POSIX::LC_ALL, 'en_CA.utf8');
  @result_std = sort @list;
  @result_ncmp = sort {ncmp($a, $b)} @list;
  # @result_std contains  A, a, B, b, C, c
  # @result_ncmp contains A, B, C, a, b, c

To be able to sort a list with an arbitrary locale it is necessary to use the C<sorted> function with a C<locale>
keyword argument:

  use Sort::Naturally::XS qw/sorted/;

  my $list = ['a.'.'c', 'A'..'B'];

  my $result_us = sorted($list, locale => 'en_US.utf8');
  # $result_us contains a, A, b, B, c, C

  my $result_ca = sorted($list, locale => 'en_CA.utf8');
  # $result_ca contains A, a, B, b, C, c

Note: due to the complexity of a cross-platform support, a locale aware sorting is guaranteed on Unix-like operating
systems only.


=head1 EXPORT

By default the module exports C<ncmp> and C<nsort> subroutines.


=head1 BENCHMARK

  require Benchmark;
  require Sort::Naturally::XS;
  require Sort::Naturally;

  my @list = (
      'H4', 'T25', 'H5', 'T27', 'H8', 'T30', 'HEX', 'T35', 'M10', 'T4', 'M12', 'T40', 'M13',
      'T45', 'M14', 'T47', 'M16', 'T5', 'M4', 'T50', 'M5', 'T55', 'M6', 'T6', 'M7', 'T60',
      'M8', 'T7', 'M9', 'T70', 'Ph0', 'T8', 'Ph1', 'T9', 'Ph2', 'TT10', 'Ph3', 'TT15', 'Ph4',
      'TT20', 'Pz0', 'TT25', 'Pz1', 'TT27', 'Pz2', 'TT30', 'Pz3', 'TT40', 'Pz4', 'TT45',
      'R10', 'TT50', 'R12', 'TT55', 'R13', 'TT6', 'R14', 'TT60', 'R5', 'TT7', 'R6', 'TT70',
      'R7', 'TT8', 'R8', 'TT9', 'S', 'TX', 'Sl', 'XZN', 'T10', 'T15', 'T20'
  );

  Benchmark::cmpthese(-3, {
      my => sub { Sort::Naturally::XS::nsort(@list) },
      other => sub { Sort::Naturally::nsort(@list) },
  });

  #          Rate other    my
  # other   561/s    --  -97%
  # my    20693/s 3588%    --

  Benchmark::cmpthese(-10, {
      std   => sub { sort @list },
      other => sub { sort {Sort::Naturally::ncmp($a, $b)} @list },
      my    => sub { sort {Sort::Naturally::XS::ncmp($a, $b)} @list },
  });

  #            Rate other   std    my
  # other 7977106/s    --   -3%   -5%
  # std   8232321/s    3%    --   -2%
  # my    8426303/s    6%    2%    --


=head1 NOTES

=over 4

=item There are differences in sorting outcomes compared with the Sort::Naturally module. Capital letters always come
before lower case letters, digits always come before letters.

  9x 14 foo fooa foolio Foolio foo12 foo12a Foo12a foo12z foo13a # Sort::Naturally
  9x 14 Foo12a Foolio foo foo12 foo12a foo12z foo13a fooa foolio # Sort::Naturally::XS

=item Due to a significant overhead it is not recommended for sorting lists consisting of letters or digits only.

=item Due to the complexity of a cross-platform support, a locale aware sorting is guaranteed on Unix-like operating
systems only.

=item Windows support added in ver. 0.7.6

=back

=head1 SEE ALSO

L<Module repository|https://github.com/CaballerosTeam/Sort-Naturally-XS>

=head1 AUTHOR

Sergey Yurzin, L<jurzin.s@gmail.com|mailto:jurzin.s@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2017 by Sergey Yurzin

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.

=cut


XS.c
/*
 * This file was generated automatically by ExtUtils::ParseXS version 3.18 from the
 * contents of XS.xs. Do not edit this file, edit XS.xs instead.
 *
 *    ANY CHANGES MADE HERE WILL BE LOST!
 *
 */

#line 1 "XS.xs"
#define PERL_NO_GET_CONTEXT
#include "locale.h"
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#include "ppport.h"

#include "src/nsort.h"

#include "const-c.inc"

static I32
S_sv_ncmp(pTHX_ SV *a, SV *b)
{
    const char *ia = (const char *) SvPVX(a);
    const char *ib = (const char *) SvPVX(b);
    return _ncmp(ia, ib, 0, 0);
}

static I32
S_sv_ncmp_reverse(pTHX_ SV *a, SV *b)
{
    const char *ia = (const char *) SvPVX(a);
    const char *ib = (const char *) SvPVX(b);
    return _ncmp(ia, ib, 1, 0);
}

static I32
S_sv_ncoll(pTHX_ SV *a, SV *b)
{
    const char *ia = (const char *) SvPVX(a);
    const char *ib = (const char *) SvPVX(b);
    return _ncmp(ia, ib, 0, 1);
}

static I32
S_sv_ncoll_reverse(pTHX_ SV *a, SV *b)
{
    const char *ia = (const char *) SvPVX(a);
    const char *ib = (const char *) SvPVX(b);
    return _ncmp(ia, ib, 1, 1);
}

#line 55 "XS.c"
#ifndef PERL_UNUSED_VAR
#  define PERL_UNUSED_VAR(var) if (0) var = var
#endif

#ifndef dVAR
#  define dVAR		dNOOP
#endif


/* This stuff is not part of the API! You have been warned. */
#ifndef PERL_VERSION_DECIMAL
#  define PERL_VERSION_DECIMAL(r,v,s) (r*1000000 + v*1000 + s)
#endif
#ifndef PERL_DECIMAL_VERSION
#  define PERL_DECIMAL_VERSION \
	  PERL_VERSION_DECIMAL(PERL_REVISION,PERL_VERSION,PERL_SUBVERSION)
#endif
#ifndef PERL_VERSION_GE
#  define PERL_VERSION_GE(r,v,s) \
	  (PERL_DECIMAL_VERSION >= PERL_VERSION_DECIMAL(r,v,s))
#endif
#ifndef PERL_VERSION_LE
#  define PERL_VERSION_LE(r,v,s) \
	  (PERL_DECIMAL_VERSION <= PERL_VERSION_DECIMAL(r,v,s))
#endif

/* XS_INTERNAL is the explicit static-linkage variant of the default
 * XS macro.
 *
 * XS_EXTERNAL is the same as XS_INTERNAL except it does not include
 * "STATIC", ie. it exports XSUB symbols. You probably don't want that
 * for anything but the BOOT XSUB.
 *
 * See XSUB.h in core!
 */


/* TODO: This might be compatible further back than 5.10.0. */
#if PERL_VERSION_GE(5, 10, 0) && PERL_VERSION_LE(5, 15, 1)
#  undef XS_EXTERNAL
#  undef XS_INTERNAL
#  if defined(__CYGWIN__) && defined(USE_DYNAMIC_LOADING)
#    define XS_EXTERNAL(name) __declspec(dllexport) XSPROTO(name)
#    define XS_INTERNAL(name) STATIC XSPROTO(name)
#  endif
#  if defined(__SYMBIAN32__)
#    define XS_EXTERNAL(name) EXPORT_C XSPROTO(name)
#    define XS_INTERNAL(name) EXPORT_C STATIC XSPROTO(name)
#  endif
#  ifndef XS_EXTERNAL
#    if defined(HASATTRIBUTE_UNUSED) && !defined(__cplusplus)
#      define XS_EXTERNAL(name) void name(pTHX_ CV* cv __attribute__unused__)
#      define XS_INTERNAL(name) STATIC void name(pTHX_ CV* cv __attribute__unused__)
#    else
#      ifdef __cplusplus
#        define XS_EXTERNAL(name) extern "C" XSPROTO(name)
#        define XS_INTERNAL(name) static XSPROTO(name)
#      else
#        define XS_EXTERNAL(name) XSPROTO(name)
#        define XS_INTERNAL(name) STATIC XSPROTO(name)
#      endif
#    endif
#  endif
#endif

/* perl >= 5.10.0 && perl <= 5.15.1 */


/* The XS_EXTERNAL macro is used for functions that must not be static
 * like the boot XSUB of a module. If perl didn't have an XS_EXTERNAL
 * macro defined, the best we can do is assume XS is the same.
 * Dito for XS_INTERNAL.
 */
#ifndef XS_EXTERNAL
#  define XS_EXTERNAL(name) XS(name)
#endif
#ifndef XS_INTERNAL
#  define XS_INTERNAL(name) XS(name)
#endif

/* Now, finally, after all this mess, we want an ExtUtils::ParseXS
 * internal macro that we're free to redefine for varying linkage due
 * to the EXPORT_XSUB_SYMBOLS XS keyword. This is internal, use
 * XS_EXTERNAL(name) or XS_INTERNAL(name) in your code if you need to!
 */

#undef XS_EUPXS
#if defined(PERL_EUPXS_ALWAYS_EXPORT)
#  define XS_EUPXS(name) XS_EXTERNAL(name)
#else
   /* default to internal */
#  define XS_EUPXS(name) XS_INTERNAL(name)
#endif

#ifndef PERL_ARGS_ASSERT_CROAK_XS_USAGE
#define PERL_ARGS_ASSERT_CROAK_XS_USAGE assert(cv); assert(params)

/* prototype to pass -Wmissing-prototypes */
STATIC void
S_croak_xs_usage(pTHX_ const CV *const cv, const char *const params);

STATIC void
S_croak_xs_usage(pTHX_ const CV *const cv, const char *const params)
{
    const GV *const gv = CvGV(cv);

    PERL_ARGS_ASSERT_CROAK_XS_USAGE;

    if (gv) {
        const char *const gvname = GvNAME(gv);
        const HV *const stash = GvSTASH(gv);
        const char *const hvname = stash ? HvNAME(stash) : NULL;

        if (hvname)
            Perl_croak(aTHX_ "Usage: %s::%s(%s)", hvname, gvname, params);
        else
            Perl_croak(aTHX_ "Usage: %s(%s)", gvname, params);
    } else {
        /* Pants. I don't think that it should be possible to get here. */
        Perl_croak(aTHX_ "Usage: CODE(0x%"UVxf")(%s)", PTR2UV(cv), params);
    }
}
#undef  PERL_ARGS_ASSERT_CROAK_XS_USAGE

#ifdef PERL_IMPLICIT_CONTEXT
#define croak_xs_usage(a,b)    S_croak_xs_usage(aTHX_ a,b)
#else
#define croak_xs_usage        S_croak_xs_usage
#endif

#endif

/* NOTE: the prototype of newXSproto() is different in versions of perls,
 * so we define a portable version of newXSproto()
 */
#ifdef newXS_flags
#define newXSproto_portable(name, c_impl, file, proto) newXS_flags(name, c_impl, file, proto, 0)
#else
#define newXSproto_portable(name, c_impl, file, proto) (PL_Sv=(SV*)newXS(name, c_impl, file), sv_setpv(PL_Sv, proto), (CV*)PL_Sv)
#endif /* !defined(newXS_flags) */

#line 197 "XS.c"

/* INCLUDE:  Including 'const-xs.inc' from 'XS.xs' */


XS_EUPXS(XS_Sort__Naturally__XS_constant); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_Sort__Naturally__XS_constant)
{
    dVAR; dXSARGS;
    if (items != 1)
       croak_xs_usage(cv,  "sv");
    PERL_UNUSED_VAR(ax); /* -Wall */
    SP -= items;
    {
#line 4 "./const-xs.inc"
#ifdef dXSTARG
	dXSTARG; /* Faster if we have it.  */
#else
	dTARGET;
#endif
	STRLEN		len;
        int		type;
	/* IV		iv;	Uncomment this if you need to return IVs */
	/* NV		nv;	Uncomment this if you need to return NVs */
	/* const char	*pv;	Uncomment this if you need to return PVs */
#line 222 "XS.c"
	SV *	sv = ST(0)
;
	const char *	s = SvPV(sv, len);
#line 18 "./const-xs.inc"
	type = constant(aTHX_ s, len);
      /* Return 1 or 2 items. First is error message, or undef if no error.
           Second, if present, is found value */
        switch (type) {
        case PERL_constant_NOTFOUND:
          sv =
	    sv_2mortal(newSVpvf("%s is not a valid Sort::Naturally::XS macro", s));
          PUSHs(sv);
          break;
        case PERL_constant_NOTDEF:
          sv = sv_2mortal(newSVpvf(
	    "Your vendor has not defined Sort::Naturally::XS macro %s, used",
				   s));
          PUSHs(sv);
          break;
	/* Uncomment this if you need to return IVs
        case PERL_constant_ISIV:
          EXTEND(SP, 1);
          PUSHs(&PL_sv_undef);
          PUSHi(iv);
          break; */
	/* Uncomment this if you need to return NOs
        case PERL_constant_ISNO:
          EXTEND(SP, 1);
          PUSHs(&PL_sv_undef);
          PUSHs(&PL_sv_no);
          break; */
	/* Uncomment this if you need to return NVs
        case PERL_constant_ISNV:
          EXTEND(SP, 1);
          PUSHs(&PL_sv_undef);
          PUSHn(nv);
          break; */
	/* Uncomment this if you need to return PVs
        case PERL_constant_ISPV:
          EXTEND(SP, 1);
          PUSHs(&PL_sv_undef);
          PUSHp(pv, strlen(pv));
          break; */
	/* Uncomment this if you need to return PVNs
        case PERL_constant_ISPVN:
          EXTEND(SP, 1);
          PUSHs(&PL_sv_undef);
          PUSHp(pv, iv);
          break; */
	/* Uncomment this if you need to return SVs
        case PERL_constant_ISSV:
          EXTEND(SP, 1);
          PUSHs(&PL_sv_undef);
          PUSHs(sv);
          break; */
	/* Uncomment this if you need to return UNDEFs
        case PERL_constant_ISUNDEF:
          break; */
	/* Uncomment this if you need to return UVs
        case PERL_constant_ISUV:
          EXTEND(SP, 1);
          PUSHs(&PL_sv_undef);
          PUSHu((UV)iv);
          break; */
	/* Uncomment this if you need to return YESs
        case PERL_constant_ISYES:
          EXTEND(SP, 1);
          PUSHs(&PL_sv_undef);
          PUSHs(&PL_sv_yes);
          break; */
        default:
          sv = sv_2mortal(newSVpvf(
	    "Unexpected return type %d while processing Sort::Naturally::XS macro %s, used",
               type, s));
          PUSHs(sv);
        }
#line 299 "XS.c"
	PUTBACK;
	return;
    }
}


/* INCLUDE: Returning to 'XS.xs' from 'const-xs.inc' */


XS_EUPXS(XS_Sort__Naturally__XS_ncmp); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_Sort__Naturally__XS_ncmp)
{
    dVAR; dXSARGS;
    if (items != 2)
       croak_xs_usage(cv,  "arg_a, arg_b");
    {
	const char *	arg_a = (const char *)SvPV_nolen(ST(0))
;
	const char *	arg_b = (const char *)SvPV_nolen(ST(1))
;
	int	RETVAL;
	dXSTARG;
#line 59 "XS.xs"
        RETVAL = _ncmp(arg_a, arg_b, 0, 0);
#line 324 "XS.c"
	XSprePUSH; PUSHi((IV)RETVAL);
    }
    XSRETURN(1);
}


XS_EUPXS(XS_Sort__Naturally__XS_nsort); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_Sort__Naturally__XS_nsort)
{
    dVAR; dXSARGS;
    PERL_UNUSED_VAR(cv); /* -W */
    {
#line 67 "XS.xs"
        if (!items) {
            XSRETURN_UNDEF;
        }
        AV * array = newAV();
        int i;
        for (i=0; i<items; i++) {
            av_push(array, ST(i));
        }
        sortsv(AvARRAY(array), items, S_sv_ncmp);
        for (i=0; i<items; i++) {
            ST(i) = av_shift(array);
        }
        av_undef(array);
        SvREFCNT_dec(array);
        XSRETURN(items);
#line 353 "XS.c"
    }
    XSRETURN(1);
}


XS_EUPXS(XS_Sort__Naturally__XS__sorted); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_Sort__Naturally__XS__sorted)
{
    dVAR; dXSARGS;
    if (items != 3)
       croak_xs_usage(cv,  "array_ref, reverse, locale");
    {
	SV *	array_ref = ST(0)
;
	int	reverse = (int)SvIV(ST(1))
;
	const char *	locale = (const char *)SvPV_nolen(ST(2))
;
#line 89 "XS.xs"
        if (!SvROK(array_ref) || SvTYPE(SvRV(array_ref)) != SVt_PVAV) {
            croak("Not an ARRAY ref");
        }
        AV * array = (AV *) SvRV(array_ref);
        int array_len = av_len(array) + 1;
        if (locale != NULL && strlen(locale)) {
            const char * old_locale = setlocale(LC_ALL, locale);
            if (reverse) {
                sortsv(AvARRAY(array), array_len, S_sv_ncoll_reverse);
            } else {
                sortsv(AvARRAY(array), array_len, S_sv_ncoll);
            }
            setlocale(LC_ALL, old_locale);
        } else {
            if (reverse) {
                sortsv(AvARRAY(array), array_len, S_sv_ncmp_reverse);
            } else {
                sortsv(AvARRAY(array), array_len, S_sv_ncmp);
            }
        }
#line 393 "XS.c"
    }
    XSRETURN_EMPTY;
}

#ifdef __cplusplus
extern "C"
#endif
XS_EXTERNAL(boot_Sort__Naturally__XS); /* prototype to pass -Wmissing-prototypes */
XS_EXTERNAL(boot_Sort__Naturally__XS)
{
    dVAR; dXSARGS;
#if (PERL_REVISION == 5 && PERL_VERSION < 9)
    char* file = __FILE__;
#else
    const char* file = __FILE__;
#endif

    PERL_UNUSED_VAR(cv); /* -W */
    PERL_UNUSED_VAR(items); /* -W */
#ifdef XS_APIVERSION_BOOTCHECK
    XS_APIVERSION_BOOTCHECK;
#endif
    XS_VERSION_BOOTCHECK;

        newXS("Sort::Naturally::XS::constant", XS_Sort__Naturally__XS_constant, file);
        newXS("Sort::Naturally::XS::ncmp", XS_Sort__Naturally__XS_ncmp, file);
        (void)newXSproto_portable("Sort::Naturally::XS::nsort", XS_Sort__Naturally__XS_nsort, file, "@");
        newXS("Sort::Naturally::XS::_sorted", XS_Sort__Naturally__XS__sorted, file);
#if (PERL_REVISION == 5 && PERL_VERSION >= 9)
  if (PL_unitcheckav)
       call_list(PL_scopestack_ix, PL_unitcheckav);
#endif
    XSRETURN_YES;
}

